include/boost/asio/windows_sspi/stream.hpp

```diff
@@ -92,10 +98,51 @@ class sspi_impl {
   CtxtHandle* m_context;
 };
 
+// TODO: Move away from this file
+template <typename NextLayer, typename ConstBufferSequence>
+struct async_write_impl {
+  async_write_impl(NextLayer& next_layer, const ConstBufferSequence& buffer, std::shared_ptr<sspi_impl> sspi)
+    : m_next_layer(next_layer)
+    , m_buffer(buffer)
+    , m_sspi(sspi)
+    , m_state(encrypt) {
+  }
+
+  template <typename Self>
+  void operator()(Self& self, boost::system::error_code ec = {}, std::size_t length = 0) {
+    boost::ignore_unused(length);
+    switch (m_state) {
```

Comment:
If you use Asio's fauxroutine you can express this code more naturally, instead of the switch statement. It will be easier to read and maintain.

Reply:
Thanks. I'll look into starting with this helps me get an understanding of what is going on though.